README
	     pIRS (profile based Illumina pair-end Reads Simulator)

Contents
========
1. Introduction
2. Program framework
3. Usage
4. Examples
5. Output file format
6. Notes

1 Introduction
==============

pIRS is a program for simulating paired-end reads from a reference genome.  It
is optimized for simulating reads similar to those generated from the Illumina
platform.

See `INSTALL' for installation instructions.

There are two subcommands: `pirs simulate' and `pirs diploid'.  See section 3
for more details, or run `pirs simulate -h' or `pirs diploid -h'.

2 Program framework
===================
  2.1 Profile Generator
    Six tools are supplied. SOAP2 or BWA, soap.coverage (http://soap.genomics.org.cn/soapaligner.html) 
  are required. The full process is shown in getprofile.sh.example as an example.
    2.1.1 GC%-Depth Profile Stat.
      a). Run soap and soap.coverage to get .depth single file(s). gzip is OK to over it.
      b). Run gc_coverage_bias on all depth single files. You will get gc-depth stat by 1 GC% and other files.
      c). Run gc_coverage_bias_plot on the gc-depth stat file. You'll get PNG plot and a .gc file by 5 GC%.
      d). Manually check the .gc file for any abnormal levels due to the lower depth on certain GC% windows.
    2.1.2 Base-Calling Profile Stat:
      a). Run soap or bwa to get .{soap,single} or .sam file(s).
      b). Run error_matrix_calculator on those file(s). You will get *.{count,ratio}.matrix .
      c). You can use error_matrix_merger to merge several .{count,ratio}.matrix files.
         However, it is up to you to keep the read length matches.
    2.1.3 InDel Profile Stat:
      a). Choose samples with NO polymorphism InDel, such as the Coliphage samples that shipped with Illumina Sequencers.
      b). Run bwa to get .sam/.bam file.
      c). Run indelstat_sam_bam to get the profile.
    2.1.4 Insert size & mapping ratio stat:
      a). Run soap or bwa to get .{soap,single} or .sam file(s).
      b). Run alignment_stator *.
         * alignment_stator cannot stat. mapping ratio for sam files now.
  2.2 Simulator
    Two commands:
    pirs diploid: use for generating diploid genome sequence. Read the input genome sequence and 
                  then simulate SNP, InDel, SV(structure variation) on it. At last, output the 
                  result genome sequence.
    pirs simulate: use for simulating Illumina data, output PE-read file.
    Note:
      a) If you only want to simulate the diploid genome sequence, "pirs diploid" is enough.
      b) If you want to simulate sequencing data of haploid genome, only you need is "pirs simulate".
      c) If you want to simulate sequencing data of diploid genome, you first need to run "pirs diploid" 
         to get the other diploid genome sequence, and then run "pirs simulate" using both the original 
         genome sequence and the previous output sequence as the input.

3 Usage
=======

 pirs <command> [option]
    diploid     generate diploid genome.
    simulate    simulate Illumina reads.

  3.1 pirs diploid
	Usage: pirs diploid [OPTIONS...] REFERENCE

	Simulate a diploid genome by creating a copy of a haploid genome with
	heterozygosity introduced.  REFERENCE specifies a FASTA file containing
	the reference genome.  It may be compressed (gzip).  It may contain multiple
	sequences (scaffolds or chromosomes), each marked with a separate FASTA tag
	line.  The introduced heterozygosity takes the form of SNPs, indels, and
	large-scale structural variation (insertions, deletions and inversions).
	If REFERENCE is '-', the reference sequence is read from stdin, but it must be
	uncompressed.

	The probabilities of SNPs, indels, and large-scale structural variation can be
	specified with the '-s', '-d', and '-v' options, respectively.  You can also
	set the ratio of transitions to transversions (for SNPs) with the '-R' option.

	Indels are split evenly between insertions and deletions. The length
	distribution of the indels is as follows and is derived from panda
	re-sequencing data:
		1bp	64.82%
		2bp	17.17%
		3bp	7.20%
		4bp	7.29%
		5bp	2.18%
		6bp	1.34%

	Large-scale structural variation is split evenly among large-scale insertions,
	deletions, and inversions.  By default, the length distribution of these
	large-scale features is as follows:
		100bp	70%
		200bp	20%
		500bp	7%
		1000bp	2%
		2000bp	1%

	`pirs diploid' does not use multiple threads, even if pIRS was configured with
	--enable-multiple threads.

	OPTIONS:
	  -s, --snp-rate=RATE    A floating-point number in the interval [0, 1] that
				   specifies the heterozygous SNP rate.  Default: 0.001

	  -d, --indel-rate=RATE  A floating-point number in the interval [0, 1] that
				   specifies the heterozygous indel rate.
				   Default: 0.0001

	  -v, --sv-rate=RATE     A floating-point number in the interval [0, 1] that
				 specifies the large-scale structural variation
				 (insertion, deletion, inversion) rate in the diploid
				   genome. Default: 0.000001

	  -R, --transition-to-transversion-ratio=RATIO
				 In a SNP, a transition is when a purine or pyrimidine
				   is changed to one of the same (A <=> G, C <=> T)
				   while a transversion is when a purine is changed
				   into a pyrimidine or vice-versa.  This option
				   specifies a floating-point number RATIO that gives
				   the ratio of the transition probability to the
				   transversion probability for simulated SNPs.
				   Default: 2.0

	  -o, --output-prefix=PREFIX
				 Use PREFIX as the prefix of the output file and logs.
				   Default: "pirs_diploid"

	  -O, --output-file=FILE
				Use FILE as the name of the output file. Use '-'
				   for standard output; this also moves the
				   informational messages from stdout to stderr.

	  -c, --output-file-type=TYPE
				 The string "text" or "gzip" to specify the type of
				   the output FASTA file containing the diploid copy
				   of the genome, as well as the log files.
				   Default: "text"

	  -n, --no-logs          Do not write the log files.

	  -S, --random-seed=SEED Use SEED as the random seed. Default:
				   time(NULL) * getpid()

	  -q, --quiet            Do not print informational messages.

	  -h, --help             Show this help and exit.

	  -V, --version          Show version information and exit.

	EXAMPLE:
	  ./pirs diploid ref_sequence.fa -s 0.001 -d 0.0001 -v 0.000001\
			 -o ref_sequence >pirs.out 2>pirs.err
      
  3.2 pirs simulate

	Usage: ./pirs simulate [OPTION]... REFERENCE.FASTA...

	pIRS is a program for simulating paired-end reads from a genome.  It is
	optimized for simulating reads from the Illumina platform.  The input to
	pIRS is any number of reference sequences. Typically you would just provide
	one FASTA file containing your reference sequence, but you may provide two
	if you have generated a diploid sequence with `pirs diploid', or if you have
	chromosomes split up into multiple FASTA files.  The output of pIRS is two
	FASTQ files containing the simulated paired-end reads, as well as several log
	files.

	Input sequences are assumed to be haploid.  If you instead want to simulate
	reads from a diploid genome, you must give the --diploid option so that
	the diploidy is taken into account when computing coverage.  If you do
	not do this, you will get twice as many reads as you wanted.

	pIRS simulates a normally-distributed insert (fragment) length using the
	Box-muller method.  Usually you want the insert length standard deviation to
	be 1/20 or 1/10 of the insert length mean (see the -m and -v options).
	This program also simulates Illumina sequencing error, quality score and
	GC bias based on empirical distribution profiles. Users may use the default
	profiles in this package, which are generated by large real sequencing data,
	or they may generate their own profiles.

	OPTIONS:
	  -l LEN, --read-len=LEN
			Generate reads having a length of LEN.  Default: 100

	  -x VAL, --coverage=VAL
			 Set the average sequencing coverage (sometimes called depth).
			 It may be either a floating-point number or an integer.

	  -m LEN, --insert-len-mean=LEN
			 Generate inserts (fragments) having an average length of LEN.
			 Default: 180

	  -v LEN, --insert-len-sd=LEN
			 Set the standard deviation of the insert (fragment) length.
			 Default: 10% of insert length mean.

	  -j, --jumping, --cyclicize
			 Make the paired-end reads face away from either other, as
			 in a jumping library.  Default: the reads face towards each
			 other.

	  -d, --diploid
			 This option asserts that reads are being simulated from a
			 diploid genome.  It causes the program to abort if there
			 are not exactly two reference sequences; in addition, the
			 coverage is divided in half, since the two reference
			 sequences are in reality the same genome.  This option
			 is not required to simulate diploid reads, but you must
			 set the coverage correctly otherwise (it will be half
			 as much as you think).

	  -B FILE, --base-calling-profile=FILE, --subst-error-profile=FILE
			 Use FILE as the base-calling profile.  This profile will be
			 used to simulate substitution errors.  Default:
			 PREFIX/share/pirs/Base-Calling_Profiles/humNew.PE100.matrix.gz

	  -I FILE, --indel-error-profile=FILE, --indel-profile=FILE
			 Use FILE as the indel-error profile.  This profile will be
			 used to simulate insertions and deletions in the reads that
			 are artifacts of the sequencing process.  Default:
			 PREFIX/share/pirs/InDel_Profiles/phixv2.InDel.matrix

	  -G FILE, --gc-bias-profile=FILE, --gc-content-bias-profile=FILE
			 Use FILE as the GC content bias profile.  This profile will
			 adjust the read coverage based on the GC content of
			 fragments.  Defaults:
			 PREFIX/share/pirs/GC-depth_Profiles/humNew.gcdep_100.dat,
			 PREFIX/share/pirs/GC-depth_Profiles/humNew.gcdep_150.dat,
			 PREFIX/share/pirs/GC-depth_Profiles/humNew.gcdep_200.dat,
			 depending on the mean insert length.

	  -e FILE, --error-rate=RATE, --subst-error-rate=RATE
			 Set the substitution error rate.  The base-calling profile
			 will still be used, but the average frequency of errors will
			 be changed to RATE.  Set to 0 to disable substitution errors
			 completely.  In that case, the base-calling profile will not
			 be used.  Default: default error rate of base-calling
			 profile.

			 Note: since pIRS parameterizes the error rate by
			 several parameters, it is very difficult to determine exactly
			 what needs to be done to make the error rate be a given
			 value.  We try to adjust the probabilities of getting each
			 quality score in order to accomodate the user-supplied error
			 rate.  However, depending on your input sequences, the actual
			 error rate simulated by pIRS could be off by 20% or more.
			 Please check the informational output to see the final error
			 rate that was actually simulated.

	  -A ALGO, --substitution-error-algorithm=ALGO, --subst-error-algo=ALGO
			 Set the algorithm used for simulating substitition errors.
			 It may be set to the string "dist" or "qtrans".  The
			 default is "qtrans".

			 The "dist" algorithm looks up the substitution error rate
			 for each base pair based on the current cycle and the true
			 base.  This lookup produces a quality score and a called base
			 that may or may not be the same as the true base.  In the
			 base-calling profile, the matrix we use is marked as the
			 [DistMatrix].

			 The "qtrans" algorithm is a Markov-chain model based on the
			 previous quality score and current cycle.  The next quality
			 score is looked up with a certain probability based on these
			 parameters.  The matrix used for this is marked as
			 [QTransMatrix] in the base-calling profile. Then, the the
			 DistMatrix is used to find a called base for the quality score.
			 The DistMatrix is also used to call the base in the first
			 cycle.

	  -M MODE, --mask=MODE, --eamss=MODE

NEWS
version 2.0

Program rewritten.

These changes are not listed in any particular order and do not include every
single change made.  Note: the format of the profile files is unchanged; the new
pIRS can read the same profile files as the old one.  Also, few changes were
made to the actual simulation algorithms.  (However, one algorithmic change was
made in how the GC content bias works; it now uses the full fragment to compute
GC content.)

- `pirs simulate' can now be compiled as a multi-threaded program.  I've
  observed speedup of 17x compared to the old version when running on a
  server-class computer.  However, compared to the new version without
  multi-threading support, the speedup is less because the new version is faster
  than the old version even if only 1 thread is used.  The exact speedup will
  depend on the computer used and the simulation parameters and inputs.  Every
  simulated read pair needs to pass through the thread running
  read_info_writer_thread_proc(), so this may be the bottleneck in the
  mulithreaded version; however, this is definitely not the bottleneck in the
  serial version, since it takes a long time to do the base calling to generate
  substitution errors and quality values.

- The random number generator now uses a 64-bit Mersenne twister algorithm.
  This may reduce bias when doing large simulations.  For example, if you are
  simulating reads from a 100 megabase sequence, you don't want to use 32-bit
  random numbers because some numbers would (for example) be 21 mod 100000000
  and other numbers would be 20 mod 100000000; you really want every position to
  be equally probable.  There are 2 versions of the 64-bit Mersenne twister
  code: the normal version, and the SSE2-accelerated version.  Both are written
  by the inventor of the algorithm.  By default, the SSE2-accelerated version is
  used; to use the normal version, provide the --disable-sse2 option to the
  configure script.

- Long command line options are now supported for both pirs simulate and pirs
  diploid.  Some options were renamed, or changed from requiring an argument to
  being a boolean option.  There are some new options, such as --quiet,
  --no-logs, and --random-seed.  Command line usage information is improved.
  The "-i" and "-I" options were removed; instead you should directly supply the
  reference sequences on the command line as non-option arguments.  `pirs
  simulate' will simulate reads from all reference sequences specified.  By
  default, reads are taken at full coverage from all sequences, so if you are
  simulating a diploid sequence, you must use half coverage, or else provide the
  --diploid option.  To help avoid confusion, this is noted in the program help,
  and a notice is printed if there were 2 input sequences but --diploid was not
  provided.

- The old version required the `pirs' binary to be invoked with an absolute
  path, which is annoying and nonstandard.  Instead, the new version installs
  the profiles to $pkgdatadir, and these paths are hard-coded into the binary.
  In other words, just run `./configure && make install' (optionally specifying
  --prefix for configure), and pIRS will find its profiles in the installed
  location.

- The default output format is now text; use -c gzip or --output-file-type=gzip
  to write gzip files.  I think it makes more sense to have the default be text.

- GC content bias is now based on the full fragment.  This may be more accurate.
  See: "Summarizing and correcting the GC content bias in high-throughput
  sequencing",
  http://nar.oxfordjournals.org/content/early/2012/02/08/nar.gks001.long

- Cumulative probability arrays that previously were using doubles are now using
  64-bit unsigned integers.  This avoids the need to convert random numbers to
  doubles before using them, and it also increases the accuracy since we can use
  2**64 different random values instead of just the doubles between 0.0 and 1.0.

- BaseCallingProfile, IndelProfile, and GCBiasProfile classes have been added to
  avoid cluttering the main code.  For example, we call a base with the
  BaseCallingProfile::call_base() function.  To decide whether to accept an
  insert based on GC content, simply call GCBiasProfile::accept_insert().

- Code duplication has been reduced.  Instead of copying code to simulate
  base-calling for each of the two reads, we simply use the same functions with
  slightly different parameters.  Similarly, instead of copying the code to
  output a read two times (one for each read), we just have an output_read()
  function that is called for both reads.

- pirs diploid now uses one random number to determine the heterozygosity type
  at a base, so it doesn't need to generate multiple random numbers per base.

- Code in pirs diploid has been de-duplicated; for example there is a function
  to try doing an indel that is called 4 times.

- pirs diploid now logs both the original position and the new position for
  SNPs, since they may differ because of indels.

- More information (such as the time and program command line) is logged to the
  beginning of each of the log files.  Some of the parameters echo echoed to
  standard output.

- `configure' script now supports conditionally omitting support for either
  `pirs simulate' or `pirs diploid' if desired.  By default, both are supported.

- InputStream and OutputStream classes were added that automatically handle
  compressed and uncompressed files, and they always exit with an error message
  if a read or write error occurs.

- Some inefficient data structures and unnecessary allocation and deallocation
  of C++ objects have been removed from the inner simulation loop.  I had
  originally changed the code to use a lot of C arrays, but then I introduced
  the `struct Read', which contains some std::vector's.  That's fine, since the
  `struct Read's are re-used and we don't have to waste time allocating and

src/pirs/simulate.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <cmath>
#include <stdint.h>
#include <map>
#include <boost/lexical_cast.hpp>
#include <boost/algorithm/string.hpp>
#include "simulate.h"

//from ASCII of A(N) C G T to 0 1 2 3, auto dealing with upper or lower case.
//8bit char type, A=a=N=n=0, C=c=1, G=g=2, T=t=3, others as 4.
char alphabet[128] =
{
 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
 4, 0, 4, 1, 4, 4, 4, 2, 4, 4, 4, 4, 4, 4, 0, 4, 
 4, 4, 4, 4, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
 4, 0, 4, 1, 4, 4, 4, 2, 4, 4, 4, 4, 4, 4, 0, 4,
 4, 4, 4, 4, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4
};


//from ASCII of A C G T to 0 1 2 3, auto dealing with upper or lower case.
//8bit char type, A=a=0, C=c=1, G=g=2, T=t=3, others as 4.
char alphabet2[128] =
{
 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
 4, 0, 4, 1, 4, 4, 4, 2, 4, 4, 4, 4, 4, 4, 4, 4, 
 4, 4, 4, 4, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
 4, 0, 4, 1, 4, 4, 4, 2, 4, 4, 4, 4, 4, 4, 4, 4,
 4, 4, 4, 4, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4
};

//from 0 1 2 3 4 to A C G T N
char Bases[5] ={
		'A', 'C', 'G', 'T', 'N'
};

//from 0 1 2 3 4 to T G C A N
char c_bases[5] ={
		'T', 'G', 'C', 'A', 'N'
};

//check whether a sequence contain non base characters, such as "N"
int check_seq (string &seq)
{       
	int is_good = 1;
	for (int i = 0; i < seq.size(); i++)
	{   if (alphabet2[seq[i]] == 4)
			{   is_good = 0;
				break;
			}
	}
	return is_good;
}

//get snp base
char get_snp_match(char base, double snp_bias[]){
	char bases[4][3]={{'G','T','C'},  //A->G transition  A->C/T transvertion
			 {'C','G','A'},  	//T->C transition  T->G/A transvertion
			 {'T','A','G'},   //C->T transition  C->A/T transvertion
			 {'A','T','C'}};  //G->A transition  G->T/C transvertion
			 	
	int a;
	char n='N';
	switch (base)
	{
  	case 'A': a=0;break;
  	case 'T': a=1;break;
  	case 'C': a=2;break;
  	case 'G': a=3;break;
  	default: return n;
	}
	
	double num = double(rand())/double(RAND_MAX);
	int i = 0;
	for(i=0; i<3; i++)
	{
		double p = snp_bias[i];
		if(num <= p){break;}
	}
	return bases[a][i];
	
}

//Getting insert sequence when heterozygous indel rate is bigger than 0
string get_insertion(int num){
	char base[]={'A','T','G','C'};
	string s;
	for (int a=0;a<num;a++)
	{
		int index=int(rand()%4);
		s.push_back(base[index]);
	}
	return s;
}

//string get_diploid_seq(string raw_seq, string id, ofstream &indel_info, ofstream &snp_info, ofstream &inversion_info, double hetersnp_rate, double heterindel_rate, double SV_rate, double snp_bias[])
string simulate_diploid_seq(string id,string raw_seq, ofstream &snp_info, ofstream &indel_info, ofstream &inversion_info, 
	double hetersnp_rate, double heterindel_rate, double SV_rate, double snp_bias[])
{
	//convert lower case to upper case 
	boost::to_upper(raw_seq);
	
	cerr<< "Begin to simulate diploid sequence..."<<endl;
	
	//probability distribution
	//small insertion and deletion share half of total heterindel_rate
	double small_insertion_rate = heterindel_rate/2.0;
	double samll_deletion_rate = heterindel_rate/2.0;
	//insertion,deletion and inversion share one third of total SV respectively
	double sv_insertion_rate = SV_rate/3.0;
	double sv_deletion_rate = SV_rate/3.0;
	double sv_inversion_rate =  SV_rate/3.0;	
	
	//indel-length 1(64.82%), 2(17.17%), 3(7.2%), 4(7.29%), 5(2.18%), 6(1.34%)
	int indel_len[6] = {1,2,3,4,5,6};
	double array1[6] = {0.6482, 0.8199, 0.8919, 0.9648, 0.9866, 1};
	
	//SV-length 100(70%),200(20%),500(7%),1000(2%),2000(1%)
	int sv_len[5] = {100, 200, 500, 1000, 2000};
	double array2[5] = {0.70, 0.90, 0.97, 0.99, 1};
	
	//stat-counter
	uint64_t snp_num = 0, small_insert_time = 0, small_insert_num = 0, small_delet_time = 0, 
		small_delet_num = 0, sv_insert_time = 0, sv_insert_num = 0, sv_delet_time = 0, sv_delet_num = 0,
		sv_invert_time = 0, sv_invert_num = 0;
	
	string diploid_seq;
	
	uint64_t seq_len = raw_seq.size();
	
	//start to simulate diploid sequence
	//simulation order: SNP -> small-deletion -> small-insertion -> SV-deltion -> SV-insertion -> SV-inversion
	for(uint64_t raw_seq_index = 0, diploid_seq_index = 0; raw_seq_index < seq_len; diploid_seq_index++, raw_seq_index++)
	{
		//////////////////simulate snp//////////////////
		double random_num = (double)rand() / double(RAND_MAX);
		if(random_num < hetersnp_rate && raw_seq[raw_seq_index] != 'N')
		{
			//get snp base
			diploid_seq += get_snp_match(raw_seq[raw_seq_index], snp_bias);
			
			snp_num++;
			uint64_t pos = raw_seq_index + 1;
			snp_info << id << "\t" << pos << "\t" << raw_seq[raw_seq_index] << "\t" << diploid_seq[diploid_seq_index] <<endl;
			
		}else{
			diploid_seq += raw_seq[raw_seq_index];
		}
	
		//////////////////simulate deletion//////////////////
		random_num = (double)rand() / double(RAND_MAX);
		if(random_num < samll_deletion_rate)
		{
			double random_num2 = (double)rand() / double(RAND_MAX);
			int i = 0;
			for(i = 0; i < 6; i++)
			{
				if(random_num2 <= array1[i])
				{
					break;
				}
			}
			if(raw_seq_index + indel_len[i] <= seq_len - 1)
			{
  			string sub_str = raw_seq.substr(raw_seq_index + 1, indel_len[i]);
  			if(check_seq(sub_str))  //check whether contain 'N' or other nonbases char
  			{ 
    			string del_seq;
    			for (int k = 1; k <= indel_len[i]; k++)
    			{
    				del_seq += raw_seq[raw_seq_index+k];
    			}
    			indel_info << id << "\t" << raw_seq_index + 1 << "\t-\t" << indel_len[i] << "\t" << del_seq <<endl;
    			raw_seq_index += indel_len[i];
    			small_delet_time++;
    			small_delet_num += indel_len[i];
  			}
			}
		}else{
			//////////////////simulate insertion//////////////////
			random_num = (double)rand() / double(RAND_MAX);
  		if(random_num < small_insertion_rate) 
  		{
  			double random_num2 = (double)rand() / double(RAND_MAX);
  			int i = 0;
  			for(i = 0; i < 6; i++)
  			{
  				if(random_num2 <= array1[i]){break;}	
  			}
  			string insert_seq = get_insertion(indel_len[i]);
  			
  			indel_info << id << "\t" << raw_seq_index + 1 << "\t+\t" << indel_len[i] << "\t" << insert_seq <<endl;
  			diploid_seq += insert_seq;
  			diploid_seq_index += indel_len[i];

  			small_insert_time++;
  			small_insert_num += indel_len[i];
  		}else{
  			//////////////////simulate sv-deletion//////////////////
  			random_num = (double)rand() / double(RAND_MAX);
        if(random_num < sv_deletion_rate) 
        {
          double random_num2 = (double)rand() / double(RAND_MAX);
          int i = 0;
          for(i = 0; i < 5; i++)
          {
          	if(random_num2 <= array2[i])
          	{
          		break;
          	}
       		}
          if(raw_seq_index + sv_len[i] <= seq_len - 1 )
          {
            string sub_str = raw_seq.substr(raw_seq_index + 1, sv_len[i]);
            if(check_seq(sub_str)) //check whether contain 'N' or other nonbases char
            {
        			string del_seq;
        			for (int k = 1; k <= sv_len[i]; k++)
        			{
        				del_seq += raw_seq[raw_seq_index+k];
        			}
        			indel_info << id << "\t" << raw_seq_index + 1 << "\t-\t" << sv_len[i] << "\t" << del_seq <<endl;
        			raw_seq_index += sv_len[i];
        			sv_delet_time++;
        			sv_delet_num += sv_len[i];
          	}
        	}
        }else{
        	//////////////////simulate sv-insertion//////////////////
  				random_num = (double)rand() / double(RAND_MAX);
      		if(random_num < sv_insertion_rate)
      		{
      			double random_num2 = (double)rand() / double(RAND_MAX);
      			int i = 0;
      			for(i = 0; i < 5; i++)
      			{
      				if(random_num2 <= array2[i]){break;}	
      			}
      			string insert_seq = get_insertion(sv_len[i]);
      			
      			indel_info << id << "\t" << raw_seq_index + 1 << "\t+\t" << sv_len[i] << "\t" << insert_seq <<endl;
      			diploid_seq += insert_seq;
      			diploid_seq_index += sv_len[i];
      			sv_insert_time++;
      			sv_insert_num += sv_len[i];
      		}else{
      			//////////////////simulate sv-inversion//////////////////
        		random_num = (double)rand() / double(RAND_MAX);
        		if(random_num < sv_inversion_rate)
        		{
        			double random_num2 = (double)rand() / double(RAND_MAX);
        			int j = 0;
        			for(j=0; j<5; j++)
        			{
        				if(random_num2 <= array2[j])
        				{
        					break;
        				}
        			}
        			if(raw_seq_index+sv_len[j] <= seq_len - 1)
        			{
            		string sub_seq = raw_seq.substr(raw_seq_index + 1, sv_len[j]);
            		if(check_seq(sub_seq)) //check whether contain 'N' or other nonbases char
            		{ 
            			inversion_info << id << "\t" << raw_seq_index + 1 << "\t" << sv_len[j] <<endl;
              		string rc_sub_seq = reversecomplementary(sub_seq);

            			diploid_seq += rc_sub_seq;
            			diploid_seq_index += sv_len[j];
            			raw_seq_index += sv_len[j];
        
              		sv_invert_time++;
              		sv_invert_num += sv_len[j];
            		}
          		}
        		}
      		}
        }
  		}
		}
	}
	
	cout<<"\n*************heterozygosis table for seq: "<<id<<"*************"<<endl
		<<"raw_seq_length\tnew_seq_length"<<endl
		<<uint64_t(raw_seq.size())<<"\t"<<uint64_t(diploid_seq.size())<<endl
		<<"heterSNP & heterINDEL:"<<endl
		<<"total_snp_num\tdeletion_time\ttotal_deletion_length\tinsertion_time\ttotal_insertion_length"<<endl
		<<snp_num<<"\t"<<small_delet_time<<"\t"<<small_delet_num<<"\t"<<small_insert_time<<"\t"<<small_insert_num<<endl
		<<"structural variation:"<<endl
		<<"deletion_time\ttotal_deletion_length\tinsertion_time\ttotal_insertion_length\tinversion_time\ttotal_inversion_length"<<endl
		<<sv_delet_time<<"\t"<<sv_delet_num<<"\t"<<sv_insert_time<<"\t"<<sv_insert_num<<"\t"<<sv_invert_time<<"\t"<<sv_invert_num<<endl;

	cerr<<"Have finished simulating seq: "<<id<<endl;
	
	return diploid_seq;
	
}

char get_base_by_Qscore(char ref_base, int qscore)
{
	double error_rate = pow(10.0, -qscore/10.0);
	double random_num = (double)rand() / double(RAND_MAX);
	char base[]={'A','T','G','C'};
	if(random_num < error_rate)
	{
		while(1)
		{
  		int index=int(rand()%4);
  		if(base[index] != ref_base)
  		{
  			return base[index];
  		}
  	}
	}else{
		return ref_base;
	}
}

//Binary search the random number location
int search_location(double *Arr, uint64_t ArrNum, double random_num){

	uint64_t left = 0;
	uint64_t right = ArrNum;
	uint64_t middle = (left+right)/2;
	
	if(random_num < Arr[0]){return 0;}
	if(random_num > Arr[ArrNum-1]){ return ArrNum;}
	
	//return the first location of bigger than 0
	if(random_num == 0){
		for(uint64_t i = 0; i < ArrNum; i++)
		{
			if(Arr[i] > 0){return i;}
		}
	}
	
	while(!(random_num > Arr[middle-1] && random_num <= Arr[middle]))
	{
		if (left == middle || right == middle){
			cerr <<"left == middle || right == middle"<<endl;
			return middle;
		}
		if(random_num > Arr[middle]){
			left = middle;
		}else if(random_num < Arr[middle-1]){
			right = middle;
		}else if(random_num == Arr[middle-1]){  //return the first region that upper boundary equal to random_num
			for(uint64_t i = middle-1; i>0; i--)
			{
				if(Arr[i-1] != Arr[i]){return i;}
			}
		}
		middle = (left+right)/2;
	}
	
	return middle;
}

//get the reverse and complement sequence
string reversecomplementary (string read)
{	
	string rc_read;
	for (int64_t i=read.size()-1; i>=0; i--)
	{	
		rc_read.push_back(c_bases[alphabet[read[i]]]);
	}
	return rc_read;
}

//simulate normal distribution insertsize by Box-muller method
int simulate_insertsize(int mean, int sd)
{
	int insertsize = 0;
	double x1,x2,radius_pow2,y1 = 0.0;
	
	do
	{
		x1 = 2.0*double(rand())/double(RAND_MAX)-1.0;
		x2 = 2.0*double(rand())/double(RAND_MAX)-1.0;
		radius_pow2 = x1*x1+x2*x2;
	}while(radius_pow2>=1.0 || radius_pow2 == 0);
	
	radius_pow2 = sqrt(-2.0*log(radius_pow2)/radius_pow2);
	y1 = x1*radius_pow2;
	insertsize = int(mean+y1*sd);
	
	return insertsize;
}


//simulate GC bias
int simulate_GC_bias(string insert_str, double *GC_bias_abundance){
	int is_ignore = 0;
	int GC_num = 0;
	int insert_size = insert_str.size();
	//get GC rate
	for (int i=0; i<insert_size; i++)
	{	
		if(insert_str[i] == 'G' || insert_str[i] == 'C'){GC_num++;}
	}
	
	double GC_rate = double(GC_num)/double(insert_size);

	//get relative abundance
	double bias_abund = GC_bias_abundance[int(GC_rate*100)];
	
	//decide whether ignore this insert string.
	double num = double(rand())/double(RAND_MAX);
	if(num > bias_abund){is_ignore = 1;}
		
	return is_ignore;
}

void get_reads_indel(int read_len, map<int,string,less<int> > &indel1, map<int,string,less<int> > &indel2, int &r1_indel_len, int &r2_indel_len, int InDel_max_len, double** InDel_matrix, int *InDel_num)
{
	//read1 indel
	int del_num = 0;
	int ins_num = 0;
	for(int i = 0; i < read_len; i++)
	{
  	double num=double(rand())/double(RAND_MAX);
		int location = search_location(InDel_matrix[i], InDel_max_len*2+1, num);
		int indel_num = InDel_num[location]; 
		if(indel_num == 0) //no indel
		{
		}else if(indel_num < 0) //deletion
		{
			string tem;
			for(int l = indel_num; l < 0; l++)
			{
				tem+="-";del_num++;
			}
			indel1[i] = tem;
		}else{//insertion
			string tem;
			for(int k = 0; k < indel_num; k++)
			{
				ins_num++;
				double num2 = double(rand())/double(RAND_MAX);
        short base=(short)ceil(num2*4);
        switch(base){
          case 1:tem+="A";break;
          case 2:tem+="C";break;
          case 3:tem+="G";break;
          case 4:tem+="T";
        }
			}
			indel1[i] = tem;
			i+=indel_num;
		}
	}
	
	r1_indel_len = ins_num - del_num;
	
	//read2 indel
	ins_num = 0;
	del_num = 0;
	for(int i = 0; i < read_len; i++)
	{
  	double num=double(rand())/double(RAND_MAX);
		int location = search_location(InDel_matrix[i+read_len], InDel_max_len*2+1, num);
		int indel_num = InDel_num[location]; 
		if(indel_num == 0) //no indel
		{
		}else if(indel_num < 0) //deletion
		{
			string tem;
			for(int l = indel_num; l < 0; l++)
			{
				tem+="-";del_num++;
			}
			indel2[i] = tem;
		}else{//insertion
			string tem;
			for(int k = 0; k < indel_num; k++)
			{
				ins_num++;
				double num2 = double(rand())/double(RAND_MAX);
        short base=(short)ceil(num2*4);
        switch(base){
          case 1:tem+="A";break;
          case 2:tem+="C";break;
          case 3:tem+="G";break;
          case 4:tem+="T";
        }
			}
			indel2[i] = tem;
			i+=indel_num;
		}
	}
	r2_indel_len = 	ins_num - del_num;
}

//Get indel read
string ref2read(string seq_ref, map<int,string,less<int> > indel, bool* is_insertion_pos)
{
	string read;
  if(indel.size()==0){
    read=seq_ref;
    return read;
  }
  read.clear();
  int k=0;
  read.push_back(seq_ref[0]);
  for(int i=1; i<seq_ref.size();){
    //cout<<i<<"\t"<<k<<endl;
    if(indel.count(k)==0){
        read.push_back(seq_ref[i]); i++; k++; 
    }
    else if(indel[k][0]=='-'){
    	i+=indel[k].size();
    	k++;
    }
    else{
      read+=indel[k];
      for(int l =0 ; l < indel[k].size(); l++)
      {
      	is_insertion_pos[k+l+1] = 1;
      } 
      k+=indel[k].size();
    }
  }
  while(indel.count(k)>0){
    read+=indel[k];
    for(int l =0 ; l < indel[k].size(); l++)
    {
    	is_insertion_pos[k+l+1] = 1;
    } 
    k+=indel[k].size();
  }
  return read;
}
